File: packages/javascript/src/LinkedList/LinkedList.js
const Node = require('./Node')

class LinkedList {
  constructor() {
    this.head = null
    this.tail = null
  }

  add(value) {
    const node = new Node(value)

    if (!this.head) {
      this.head = node
    } else {
      this.tail.next = node
    }

    this.tail = node
  }

  valueAt(index) {
    return this.nodeAt(index).value
  }

  nodeAt(index) {
    const length = this.size()

    if (index > length || index < 0) {
      throw new Error('Index out of bounds.')
    }

    let node = this.head

    for (let i = 1; i < index; i++) {
      node = node.next
    }

    return node
  }

  size() {
    let length = 0
    let node = this.head

    while (node) {
      length++
      node = node.next
    }

    return length
  }

  empty() {
    return this.size() === 0
  }
}

module.exports = LinkedList

File: ruby/lib/linked_list.rb
require_relative './node'

class LinkedList
  def initialize()
    @head = nil
    @tail = nil
  end

  def add(value)
    node = Node.new(value)

    if @head == nil
      @head = node
    else
      @tail.next = node
    end

    @tail = node
  end

  def value_at(index)
    node_at(index).value
  end

  def node_at(index)
    node = @head

    for i in (1..size)
      if i == index
        break
      end

      node = node.next
    end

    node
  end

  def empty
    size == 0
  end

  def size
    length = 0
    node = @head

    while node != nil
      length += 1
      node = node.next
    end

    length
  end
end
